src/files/BinaryFile.cpp

```diff
@@ -367,402 +403,226 @@ inline uint64_t swap_u64(uint64_t value) {
-    uint64_t value;
-    std::memcpy(&value, buffer, sizeof(uint64_t));
-#if CHEMFILES_BYTE_ORDER == CHEMFILES_LITTLE_ENDIAN
-    value = swap_u64(value);
-#endif
-    return value;
-}
-
-inline void u16_to_big_endian(uint16_t value, char buffer[2]) {
-#if CHEMFILES_BYTE_ORDER == CHEMFILES_LITTLE_ENDIAN
-    value = swap_u16(value);
-#endif
-    std::memcpy(buffer, &value, sizeof(uint16_t));
-}
-
-inline void u32_to_big_endian(uint32_t value, char buffer[4]) {
+template<typename T>
+inline void BigEndianFile::read_as_big_endian(T* data, size_t count) {
+    const size_t byte_count = sizeof(T) * count;
+    auto char_data = reinterpret_cast<char*>(data);
+    this->read_char(char_data, byte_count);
 #if CHEMFILES_BYTE_ORDER == CHEMFILES_LITTLE_ENDIAN
-    value = swap_u32(value);
+    auto uspan = span<T>(reinterpret_cast<T*>(data), count);
+    for (size_t i = 0; i < count; ++i) {
+        uspan[i] = swap_endianness<T>(uspan[i]);
+    }
 #endif
-    std::memcpy(buffer, &value, sizeof(uint32_t));
 }
 
-inline void u64_to_big_endian(uint64_t value, char buffer[8]) {
+template<typename T>
+inline void BigEndianFile::write_as_big_endian(const T* data, size_t count) {
+    const size_t byte_count = sizeof(T) * count;
 #if CHEMFILES_BYTE_ORDER == CHEMFILES_LITTLE_ENDIAN
-    value = swap_u64(value);
+    swap_buf_.resize(byte_count);
+    std::memcpy(swap_buf_.data(), data, byte_count);
+    auto uspan = span<T>(reinterpret_cast<T*>(swap_buf_.data()), count);
+    for (size_t i = 0; i < count; ++i) {
+        uspan[i] = swap_endianness<T>(uspan[i]);
+    }
+    this->write_char(swap_buf_.data(), byte_count);
+#else
+    this->write_char(reinterpret_cast<const char*>(data), byte_count);
 #endif
-    std::memcpy(buffer, &value, sizeof(uint64_t));
 }
 
 void BigEndianFile::read_i16(int16_t* data, size_t count) {
-    auto char_data = reinterpret_cast<char*>(data);
-    this->read_char(char_data, count * sizeof(int16_t));
-    for (size_t i=0; i<count; i++) {
-        auto value = u16_from_big_endian(char_data + sizeof(int16_t) * i);
-        std::memcpy(data + i, &value, sizeof(int16_t));
-    }
+    const auto uptr = reinterpret_cast<uint16_t*>(data);
+    read_as_big_endian<uint16_t>(uptr, count);
 }
 
 
 void BigEndianFile::read_u16(uint16_t* data, size_t count) {
-    auto char_data = reinterpret_cast<char*>(data);
-    this->read_char(char_data, count * sizeof(uint16_t));
-    for (size_t i=0; i<count; i++) {
-        data[i]  = u16_from_big_endian(char_data + sizeof(uint16_t) * i);
-    }
+    read_as_big_endian<uint16_t>(data, count);
 }
 
 
 void BigEndianFile::read_i32(int32_t* data, size_t count) {
-    auto char_data = reinterpret_cast<char*>(data);
-    this->read_char(char_data, count * sizeof(int32_t));
-    for (size_t i=0; i<count; i++) {
-        auto value = u32_from_big_endian(char_data + sizeof(int32_t) * i);
-        std::memcpy(data + i, &value, sizeof(int32_t));
-    }
+    const auto uptr = reinterpret_cast<uint32_t*>(data);
```

Comment:
I'm not sure casting from `int32_t` to `uint32_t` is valid, I'll need to check the exact rules w.r.t. type punning and aliasing.

A possible alternative would be to define `swap_endianness` for signed integers, converting them to unsigned integer with a `memcpy` on char data (this is valid and should be optimized by the compiler without issue). 

This should allow calling `read_as_big_endian<int32_t>(uptr, count);` directly here

Reply:
I had the assumption that it was permitted to pun between signed/unsigned types.
But, punning between float/double and `uintXX_t` (as I do later on) is most definitely not allowed. I'll have to look for a better way anyhow.